I2C and BLE Selective begin

giulcioffi opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
This PR allows to let the user select what to use by the time the `BHY2.begin()` function is called. This allows to avoid starting I2C or BLE threads if they are not needed.

Two options have been provided to configure the settings:

SOLUTION N.1:
`bool begin(NiclaConfig config = NICLA_BLE_AND_I2C, NiclaWiring niclaConnection = NICLA_VIA_ESLOV);`

Possible configurations:
- Both I2C and BLE enabled, Nicla connected via Eslov cable: `BHY2.begin();`
- Both I2C and BLE enabled, Nicla connected as a Shield: `BHY2.begin(NICLA_BLE_AND_I2C, NICLA_AS_SHIELD);`
- Only I2C enabled, Nicla connected via Eslov cable: `BHY2.begin(NICLA_I2C);`
- Only I2C enabled, Nicla connected as a Shield: `BHY2.begin(NICLA_I2C, NICLA_AS_SHIELD);`
- Only BLE enabled: `BHY2.begin(NICLA_BLE);`


SOLUTION N.2:
Use **NiclaSettings** class to set Nicla configuration. This method is similar to the SPISettings one, used to set SPI parameters.
`bool begin(NiclaSettings& settings);`

Possible configurations:
- Both I2C and BLE enabled, Nicla connected via Eslov cable: `NiclaSettings settings(NICLA_I2C, NICLA_BLE, NICLA_VIA_ESLOV, 0);`
- Both I2C and BLE enabled, Nicla connected as a Shield: `NiclaSettings settings(NICLA_I2C, NICLA_BLE, 0, NICLA_AS_SHIELD);`
- Only I2C enabled, Nicla connected via Eslov cable: `NiclaSettings settings(NICLA_I2C, 0, NICLA_VIA_ESLOV, 0);`
- Only I2C enabled, Nicla connected as a Shield: `NiclaSettings settings(NICLA_I2C, 0, 0, NICLA_AS_SHIELD);`
- Only BLE enabled: `NiclaSettings settings(0, NICLA_BLE, 0, 0);`
And `BHY2.begin(settings);`


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Breaking change

facchinm commented on Arduino_BHY2/src/Arduino_BHY2.h:
```diff
@@ -58,6 +78,9 @@ class Arduino_BHY2 {
   PinName _eslovIntPin;
+
+  NiclaConfig _niclaConfig;
+  NiclaSettings _settings = NiclaSettings(NICLA_I2C, NICLA_BLE, NICLA_VIA_ESLOV, 0);
```
Can this line be removed?

facchinm commented on Arduino_BHY2/src/Arduino_BHY2.cpp:
```diff
@@ -93,32 +98,58 @@ bool Arduino_BHY2::begin(NiclaWiring niclaConnection)
+bool Arduino_BHY2::begin(NiclaSettings& settings)
+{
+  uint8_t niclaSettings = settings.getConfiguration();
+  if (niclaSettings & NICLA_I2C & NICLA_BLE) {
```
Not sure this line have the expected result, should it be 
```if ((niclaSettings & NICLA_I2C) && (niclaSettings & NICLA_BLE))``` ?

facchinm reviewed (changes requested):
Apart from the two nitpicks the PR is very good :)
I'd just change the NiclaSettings constructoir signature to allow freeform settings

facchinm commented on Arduino_BHY2/src/Arduino_BHY2.h:
```diff
@@ -15,8 +15,27 @@
+
+class NiclaSettings {
+public:
+  NiclaSettings(uint8_t i2c, uint8_t ble, uint8_t eslov, uint8_t shield) {
```
If we change the signature to 
```
NiclaSettings(uint8_t conf1 = 0, uint8_t conf2 = 0, uint8_t conf3 = 0, uint8_t conf4 = 0)
```
we can go freeform with the constructor, since the defines are a bitmask, like
```
NiclaSettings settings(NICLA_I2C, NICLA_BLE, NICLA_VIA_ESLOV)
```
or 
```
NiclaSettings settings(NICLA_BLE);
```

giulcioffi commented:
All the mentioned issue should be now solved @facchinm :) I squashed the fixes in the first commit, in order to keep it atomic. 
I also added a new commit to the PR. While testing with Portenta as host, I noticed that sometimes `Wire.endTransmission()` was failing. Then I added a check to the return value of `Wire.endTransmission(`): it checks if the function returns correctly, otherwise repeats the transfer (max 3 times). This change has been tested with a Portenta and with a MKR board.
